File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let rec steps' count =
    function
    | n when n < 1 -> None
    | 1 -> Some count
    | n when n % 2 = 0 -> steps' (count + 1) (n / 2)
    | n -> steps' (count + 1) (3 * n + 1)

let steps number = steps' 0 number

File: Exercism/csharp/collatz-conjecture/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number, int count = 0) => number switch
    {
        < 1                   => throw new ArgumentOutOfRangeException(),
        1                     => count,
        var n when n % 2 == 0 => Steps(number / 2, count + 1),
        _                     => Steps((number * 3) + 1, count + 1)
    };
}